apm-agent-plugins/apm-cassandra/apm-cassandra3-plugin/src/main/java/co/elastic/apm/agent/cassandra3/Cassandra3Instrumentation.java

```diff
@@ -0,0 +1,185 @@
+        public static Object onEnter(@Advice.Argument(0) Statement statement) {
+            return cassandraHelper.startCassandraSpan(getQuery(statement), statement instanceof BoundStatement);
+        }
+
+        @Nullable
+        private static String getQuery(Statement statement) {
+            if (statement instanceof SimpleStatement) {
+                return ((SimpleStatement) statement).getQueryString();
+            } else if (statement instanceof BoundStatement) {
+                return ((BoundStatement) statement).preparedStatement().getQueryString();
+            }
+            return null;
+        }
+
+        @Advice.OnMethodExit(suppress = Throwable.class, onThrowable = Throwable.class, inline = false)
+        public static void onExit(@Advice.Thrown @Nullable Throwable thrown,
+                                  @Advice.Return ResultSetFuture result,
+                                  @Nullable @Advice.Enter Object spanObj) {
+            if (!(spanObj instanceof Span)) {
+                return;
+            }
+            final Span span = (Span) spanObj;
+            span.captureException(thrown).deactivate();
+            Futures.addCallback(result, new FutureCallback<ResultSet>() {
+                @Override
+                public void onSuccess(@Nullable ResultSet result) {
+                    if (result != null) {
+                        Host host = result.getExecutionInfo().getQueriedHost();
+                        if (host != null) {
+                            DestinationAddressSetter.Resolver.get().setDestination(span, host);
+                        }
+                    }
+                    span.end();
+                }
+
+                @Override
+                public void onFailure(Throwable t) {
+                    span.endExceptionally(t);
+                }
+            });
+        }
+    }
+
+    private interface DestinationAddressSetter {
+        void setDestination(Span span, Host host);
+
+        class Resolver {
+
+            @Nullable
+            private static volatile DestinationAddressSetter delegate;
+
+            static DestinationAddressSetter get() {
+                DestinationAddressSetter localDelegate = delegate;
+                if (localDelegate == null) {
+                    synchronized (Resolver.class) {
+                        localDelegate = delegate;
+                        if (localDelegate == null) {
+                            try {
+                                Class.forName("com.datastax.driver.core.Host").getMethod("getSocketAddress");
+                                delegate = localDelegate = (DestinationAddressSetter) Class.forName(DestinationAddressSetter.class.getName() + "$WithSocketAddress").getEnumConstants()[0];
+                            } catch (ReflectiveOperationException | LinkageError ignore) {
+                                delegate = localDelegate = WithInetAddress.INSTANCE;
+                            }
```

Comment:
That looks good!
Do you have some test app where this can be tested with class loading that better resembles real scenarios?

Reply:
I have manually tested with a standalone Spring Boot application:

<details>
<summary>Click to expand</summary>

```java
@SpringBootApplication
public class CassandraDemoApplication {

    public static void main(String[] args) {
        SpringApplication.run(CassandraDemoApplication.class, args);
    }

    @Bean
    public Cluster cluster(@Value("${cassandra.port}") int cassandraPort) {
        Cluster cluster = Cluster.builder()
                .addContactPoint("localhost")
                .withPort(cassandraPort)
                .withoutMetrics()
                .build();
        try (Session s = cluster.connect()) {
            s.execute("CREATE KEYSPACE IF NOT EXISTS test WITH replication = {'class':'SimpleStrategy','replication_factor':'1'};");
        }
        try (Session s = cluster.connect("test")) {
            s.execute("CREATE TABLE users (id UUID, name text, PRIMARY KEY(name, id))");
            s.execute(s.prepare("INSERT INTO users (id, name) values (?, ?)").bind(UUID.randomUUID(), "alice"));
        } catch (Exception ignore) {
        }
        return cluster;
    }


    @RestController
    public static class HomeController {

        private final Cluster cluster;

        public HomeController(@Autowired Cluster cluster) {
            this.cluster = cluster;
        }

        @GetMapping("/")
        public String home() {
            return "ok";
        }

        @GetMapping("/cassy")
        public String cassandra() {
            try (Session s = cluster.connect("test")) {
                return s.execute("SELECT * FROM users where name = 'alice' ALLOW FILTERING").one().getString("name");
            }
        }
    }
}
```
</details>

With version 2.0.1, it only records `destination.address`, and version 2.0.2 also records `destination.port`.